### PR TITLE
Replace .at() calls with custom method showing actual source location on failure

### PR DESCRIPTION
**Pull request by Nekotekina**
Adds a supplement for common STL .at() methods but which also provides source location.

### CODE REVIEW

File: rpcs3/util/types.hpp [hunk 1]
```diff
@@ -1056,6 +1056,41 @@ template <typename T, usz Size>
 	return static_cast<u32>(Size);
 }
 
+template <typename T, usz Size, typename U>
+[[nodiscard]] constexpr auto& at32(const T (&array)[Size], u32 index,
```
Comment (elad335) [inline rpcs3/util/types.hpp:1060]: Maybe use std::size.

Comment (elad335) [reply rpcs3/util/types.hpp:1060]: I mean this overload is better removed in this widely used header, and std::size can help with that. There's concept test in serializtion.hpp for detection of arrays, even worjs for std::span etc.

Comment (Nekotekina) [reply rpcs3/util/types.hpp:1060]: Okay, I'll look into it

File: rpcs3/Emu/RSX/Common/ranged_map.hpp [hunk 2]
```diff
@@ -177,63 +177,63 @@ namespace rsx
...
 		}
 
-		inline void emplace(const utils::address_range& range, T&& value)
+		void emplace(const utils::address_range& range, T&& value)
 		{
 			broadcast_insert(range);
 			m_data[block_for(range.start)].insert_or_assign(range.start, std::forward<T>(value));
 		}
 
-		inline iterator find(const u32 key)
```
Comment (elad335) [inline rpcs3/Emu/RSX/Common/ranged_map.hpp:186]: Seems like the methods `find` and `at` have been swapped unintendedly.

**Comment by Megamouse** [comment]
I'd rather have these data types wrapped or something instead of replacing all the calls with some obscure syntax

File: rpcs3/Emu/RSX/Common/ranged_map.hpp (hunk 2 above)
Comment (Nekotekina) [reply rpcs3/Emu/RSX/Common/ranged_map.hpp:186]: I removed `at` and added `count` to make it compatible with at32 requirements

**Review by kd-11 — APPROVED** [review]
RSX side lgtm

File: rpcs3/util/types.hpp [hunk 3]
```diff
@@ -1038,22 +1038,45 @@ template <typename To = void, typename From, typename = decltype(static_cast<To>
...
+	const u32 idx = ::narrow<u32>(+index, line, 10001, file, func);
+	const u32 csz = ::size32(container, line, 10002, file, func);
+	if (csz <= idx) [[unlikely]]
+		fmt::raw_verify_error({line, col, file, func}, u8"Out of range");
+	auto it = std::begin(std::forward<CT>(container));
+	std::advance(it, idx);
+	return *it;
+}
+
+template <typename CT, typename T> requires requires (CT&& x, T&& y) { x.count(y); x.find(y); }
```
Comment (kd-11) [inline rpcs3/util/types.hpp:1068]: Why the requirement for count()?
Don't we just need find()?

Comment (Nekotekina) [reply rpcs3/util/types.hpp:1068]: It's present in all associative containers, but I'm not really sure.

**Comment by Nekotekina** [comment]
I think an alternative (replacing container types) would be too intrusive and not always possible.

**Pull request merged**